src/odemis/driver/npmc.py

```diff
@@ -635,6 +712,51 @@ def _createMoveFuture(self):
         f.task_canceller = self._cancelCurrentMove
         return f
 
+    @isasync
+    def reference(self, axes):
+        if not axes:
+            return model.InstantaneousFuture()
+        f = self._createMoveFuture()
+        f = self._executor.submitf(f, self._doReference, f, axes)
+        return f
+
+    def _doReference(self, future, axes):
+        """
+        Actually runs the referencing code
+        axes (set of str)
+        raise:
+            IOError: if referencing failed due to hardware
+            CancelledError if was cancelled
+        """
+        # Reset reference so that if it fails, it states the axes are not
+        # referenced (anymore)
+        with future._moving_lock:
+            try:
+                # do the referencing for each axis sequentially
+                # (because each referencing is synchronous)
+                for a in axes:
+                    if future._must_stop.is_set():
+                        raise CancelledError()
+                    aid = self._axis_map[a]
+                    self.referenced._value[a] = False
+                    self.HomeSearch(aid, REF_NEGATIVE_LIMIT)  # search for the negative limit signal to set an origin
+                    self._waitEndMove(future, (aid,), time.time() + 100)  # block until it's over
+                    self.SetHome(aid, 0.0)  # set negative limit as origin
+                    self.referenced._value[a] = True
+            except CancelledError as ex:
+                logging.info("Referencing cancelled: %s", ex)
```

Comment:
At least put a big fat warning that referencing must be run again. Ideally, convince the hardware that it can already move again.